sxt/multiexp/bucket_method2/sum.h

```diff
@@ -0,0 +1,130 @@
+CUDA_CALLABLE void sum_bucket(T* __restrict__ sums, const T* __restrict__ generators,
+                              const uint16_t* __restrict__ bucket_prefix_counts,
+                              const uint16_t* __restrict__ indexes, unsigned num_buckets_per_digit,
+                              unsigned n, unsigned index) noexcept {
+  auto digit_index = index / num_buckets_per_digit;
+  auto bucket_offset = index % num_buckets_per_digit;
+
+  // adjust pointers
+  auto& sum = sums[index];
+  bucket_prefix_counts += digit_index * num_buckets_per_digit;
+  indexes += digit_index * n;
+
+  // sum
+  uint16_t first;
+  if (bucket_offset == 0) {
+    first = 0;
+  } else {
+    first = bucket_prefix_counts[bucket_offset - 1u];
+  }
+  auto last = bucket_prefix_counts[bucket_offset];
+  if (first == last) {
+    sum = T::identity();
+    return;
+  }
+  T e = generators[indexes[first++]];
+  for (; first != last; ++first) {
+    auto t = generators[indexes[first]];
+    add_inplace(e, t);
+  }
+  sum = e;
+}
+
+//--------------------------------------------------------------------------------------------------
+// sum_buckets
+//--------------------------------------------------------------------------------------------------
+/**
+ * Sum generators into buckets.
+ *
+ * This function corresponds roughly to the 1st loop of Algorithm 1 described in
+ *
+ *    PipeMSM: Hardware Acceleration for Multi-Scalar Multiplication
+ *    https://eprint.iacr.org/2022/999.pdf
+ */
+template <bascrv::element T>
+xena::future<> sum_buckets(basct::span<T> sums, basct::cspan<T> generators,
+                           basct::cspan<const uint8_t*> exponents, unsigned element_num_bytes,
+                           unsigned bit_width) noexcept {
+  auto num_buckets_per_digit = (1u << bit_width) - 1u;
+  auto num_digits = basn::divide_up(element_num_bytes * 8u, bit_width);
+  auto num_outputs = static_cast<unsigned>(exponents.size());
+  auto num_buckets_total = static_cast<unsigned>(sums.size());
+  auto n = static_cast<unsigned>(generators.size());
+  SXT_DEBUG_ASSERT(basdv::is_active_device_pointer(sums.data()));
+
+  // compute multiproduct table
+  memmg::managed_array<uint16_t> bucket_prefix_counts{num_buckets_total,
+                                                      memr::get_device_resource()};
+  memmg::managed_array<uint16_t> indexes{n * num_digits * num_outputs, memr::get_device_resource()};
+  auto fut = make_multiproduct_table(bucket_prefix_counts, indexes, exponents, element_num_bytes,
+                                     bit_width, n);
+
+  // copy generators to device
+  basdv::stream stream;
+  memr::async_device_resource resource{stream};
+  memmg::managed_array<T> generators_dev{n, &resource};
+  basdv::async_copy_host_to_device(generators_dev, generators, stream);
+
+  // sum buckets
+  memmg::managed_array<T> sums_dev{num_buckets_total, &resource};
+  co_await std::move(fut);
+  basl::info("summing {} buckets", num_buckets_total);
```

Comment:
Would `basl::info` calls like this flood any info logs? I know we use this logging elsewhere in the code. Do you know if this logging impacts performance?

Reply:
I wouldn't expect it to impact performance much. You can turn off logging and cost of logging is small relative to the computation here.